janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/AdjacentVertexFilterOptimizerStrategy.java

```diff
@@ -32,73 +32,161 @@
-                    }
+            });
+    }
+
+    private void replaceStep(Traversal.Admin<?, ?> traversal, OptimizableQueryType type,
+                             TraversalFilterStep originalStep, List<Step> steps) {
+        // Get the direction in which we filter on the adjacent vertex (or null if not a valid
+        // adjacency filter)
+        Direction direction = parseDirection(steps);
+        P predicate = parsePredicate(type, steps);
+        // Check that we have a valid direction and a valid vertex filter predicate
+        if (direction != null && isValidPredicate(type, predicate) &&
+            isPreviousStepValid(originalStep, direction)) {
+            // Now replace the step with a has condition
+            HasContainer hc =
+                new HasContainer(ImplicitKey.ADJACENT_ID.name(), P.eq(predicate.getValue()));
+            TraversalHelper.replaceStep(originalStep, new HasStep(traversal, hc), traversal);
+        }
+    }
+
+    private OptimizableQueryType analyzeSubSteps(List<Step> steps) {
+        if (steps.size() != 2) {
+            return OptimizableQueryType.NONE;
+        }
+
+        boolean validFirstStep = (steps.get(0) instanceof EdgeVertexStep);
+        validFirstStep = validFirstStep || (steps.get(0) instanceof EdgeOtherVertexStep);
 
+        if (!validFirstStep) {
+            return OptimizableQueryType.NONE;
+        }
+
+        if (steps.get(1) instanceof IsStep) {
+            // Check if this filter traversal matches the pattern: _.inV/outV/otherV.is(x)
+            return OptimizableQueryType.IS;
+        } else if (steps.get(1) instanceof HasStep) {
+            // Check if this filter traversal matches the pattern: _.inV/outV/otherV.hasId(x)
+            HasStep hasStep = (HasStep) steps.get(1);
+            if (hasStep.getHasContainers().size() != 1) {
+                // TODO does it make sense to allow steps with > 1 containers here?
+                return OptimizableQueryType.NONE;
+            }
+
+            HasContainer has = (HasContainer) hasStep.getHasContainers().get(0);
+            if (has.getKey().equals(T.id.getAccessor())) {
+                return OptimizableQueryType.HASID;
+            } else {
+                return OptimizableQueryType.NONE;
+            }
+        } else {
+            return OptimizableQueryType.NONE;
+        }
+    }
+
+    private boolean isPreviousStepValid(TraversalFilterStep originalStep, Direction direction) {
+        // check that this step is preceded by VertexStep that returns edges
+        Step<?, ?> previousStep = originalStep.getPreviousStep();
+        while (previousStep != EmptyStep.instance()) {
+            if (!(previousStep instanceof HasStep) && !(previousStep instanceof IdentityStep)) {
+                break;
+            }
+            previousStep = previousStep.getPreviousStep();
+        }
+
+        if (previousStep instanceof VertexStep) {
+            try (VertexStep<?> vertexStep = (VertexStep<?>) previousStep) {
```

Comment:
I didn't find `vertexStep.close()` in the previous version. I don't know if we should close it here after `try with resources` but could you explain why you are closing `vertexStep` here? Is it connected to resources leak or other possible problems?

Reply:
When not closing `vertexStep`, Java outputs a warning about a possible resource leak.

Reply:
@rngcntr Can you confirm a possible resource leak here?
I am just not sure if there is a resource leak. What about other possible `VertexStep` which are not connected to this optimization? Do we have a resource leak there?
What about all other steps which implement `AutoCloseable` (most of the steps are either implementing it through `TraversalParent` or directly)? Do we have resource leak there or is this resource leak connected only to this optimization?
If the resource leak is connected to all steps, then we should close them somewhere else (no in particular optimization step). If the resource leak is directly connected to this optimization, then you are doing it right and we should close those steps here.
I am not very familiar with that code base, so I don't currently see resource leak here.
What I am thinking is that maybe we should close `originalStep` because we replace it with a new step but I am not sure about it yet.
Also, I don't want to introduce a bug here in case this vertex step is supossed to be closed at some other stage but we are closing it erlier, in such case we could introduce a new bug. I don't tell that this code introduces a new bug I just tell that I don't see this code to fix any existing bug (at least, I can't verify it yet).
I think maybe it is better to not mix the possible resource leak (which isn't connected to this PR) and `hasId` optimization.
I think a separate PR to fix a possible resource leak would be a better choice (if there is a resource leak).
If you can confirm resource leak, then I am OK with making this change in this PR. If you are able to add a test which verifies this resource leak, it would be great.

Reply:
I can neither confirm the resource leak nor find any find any bugs (covered by the tests) that arise from closing the step here.
Internally, `close()` causes the underlying `FlatMapStep`'s Iterator to be closed _if it it implements_ `AutoCloseable`. This Iterator is initialized with a singleton instance of the class `EmptyIterator` which is replaced with an actual Iterator of the Elements emitted by the call of `flatMap()`.

Because `EmptyIterator` does not implement `AutoCloseable`, the call of `close()` has no other effect than satisfying the compiler ;)

I quickly discussed this with @farodin91 yesterday and we came to the conclusion that even if there's no point in calling `close()` in this particular case, it aids consistency of the source code in general.

Don't get me wrong here, I'm also fine with keeping this step unclosed if others prefer

Reply:
What do you mean by `it aids consistency of the source code in general.`?
As for me, if we are calling `close` for some steps (without any logic behind this except removing the warning from compiler) and not calling `close` for other steps then it looks like inconsistent code.
Let's imagine the situation when `close` method actually making the step unreadable. In such case, we will experience a situation, that sometimes, some of `VertexStep`s are closed during the optimization phase, but some of `VertexStep`s are not closed during the same optimization phase. We will need to track it back to this point to find out that we are closing only some `VertexStep`s here for no reason actually (besides making the compiler happy).
I've described a really easy possible bug here, which I believe we will be able to find very quickly, but in the future `close` method may have a totally different logic, which will cause "something" and so, it will be hard to find out the bug which isn't easily reproducible.
As for me, adding such logic just to make a compiler happy is a bad idea.
I believe, this change isn't connected to this optimization directly. I would recommend opening a new issue and discussing this change there, because this may take some time before we have a consensus on calling `close` here (which we are not calling right now but your PR adds this logic for some reason. As for me, the reason isn't satisfying, but other members may have more thoughts on this).

Reply:
I totally get your point concerning future changes. I reverted this change and I'm now keeping the handling of `close` calls the same as it was before this PR.